### PR TITLE
Range modifications

### PR DESCRIPTION
**Pull request by smichr**

### CODE REVIEW

File: sympy/sets/fancysets.py [hunk 1]
```diff
@@ -308,16 +308,13 @@ def _intersect(self, other):
 
 class Range(Set):
     """
-    Represents a range of integers.
+    Represents a range of 1 or more integers.
```
Comment (smichr) [inline sympy/sets/fancysets.py]: establishes that a FIniteSet is not returned for range of length 1

File: sympy/sets/fancysets.py [hunk 2]
```diff
@@ -384,90 +390,83 @@ def __new__(cls, *args):
...
-            return S.EmptySet
 
-        # normalize args: regardless of how they are entered they will show
-        # canonically as Range(inf, sup + 1, step) with step > 0 or Range(sup,
-        # inf + 1, step) with step < 0
-        if n.is_finite:
-            stop = start + n*step
-
-        if (start, stop) == (S.NegativeInfinity, S.Infinity):
+        if all(i.is_infinite for i in  (start, stop)):
```
Comment (smichr) [inline sympy/sets/fancysets.py]: this was a corner case missed and a test was added

File: sympy/sets/fancysets.py [hunk 3]
```diff
@@ -478,12 +477,12 @@ def _intersect(self, other):
         return None
 
     def _contains(self, other):
-        if (((self.start - other)/self.step).is_integer or
-            ((self.stop - other)/self.step).is_integer):
-            return _sympify(other >= self.inf and other <= self.sup)
-        elif (((self.start - other)/self.step).is_integer is False and
-            ((self.stop - other)/self.step).is_integer is False):
+        if not other.is_Integer:
```
Comment (smichr) [inline sympy/sets/fancysets.py]: TODO: I guess checking `is_integer` as True or False would be more robust

File: sympy/sets/fancysets.py [hunk 4]
```diff
@@ -501,6 +500,8 @@ def __iter__(self):
 
 
     def __len__(self):
+        if self.start.is_infinite or self.stop.is_infinite:
```
Comment (smichr) [inline sympy/sets/fancysets.py]: corner case...don't remember if I added a test

File: sympy/sets/fancysets.py [hunk 5]
```diff
@@ -510,13 +511,13 @@ def __nonzero__(self):
     __bool__ = __nonzero__
 
     def _ith_element(self, i):
+        if self.start.is_infinite:
```
Comment (smichr) [inline sympy/sets/fancysets.py]: 2 lines not needed since inf will consume the addend

File: sympy/sets/tests/test_fancysets.py [hunk 6]
```diff
@@ -186,43 +212,81 @@ def test_range_interval_intersection():
...
-
-        ]:
-        assert r.intersect(i) == result, "%s.intersect(%s) == %s" % (r, i,
-        result)
-        if r.is_finite and result.is_finite:
-            assert FiniteSet(*r).intersect(i) == FiniteSet(*result), "FiniteSet(*%s).intersect(%s) == FiniteSet(*%s)" % (r, i, result)
+
+        ]):
+        assert r
+        for rev in range(2):
```
Comment (smichr) [inline sympy/sets/tests/test_fancysets.py]: handle all reversals with this

File: sympy/sets/fancysets.py (hunk 4 above)
Comment (smichr) [inline sympy/sets/fancysets.py]: nope -- need a test

File: sympy/sets/fancysets.py (hunk 3 above)
Comment (smichr) [inline sympy/sets/fancysets.py]: done

File: sympy/sets/fancysets.py (hunk 4 above)
Comment (smichr) [inline sympy/sets/fancysets.py]: done

File: sympy/sets/fancysets.py (hunk 5 above)
Comment (smichr) [inline sympy/sets/fancysets.py]: removed

File: sympy/sets/tests/test_fancysets.py [hunk 7]
```diff
@@ -186,43 +212,81 @@ def test_range_interval_intersection():
...
+                assert not result
+            else:
+                if len(ans) == 1:
+                    ans.append(ans[0])
+                A, B = ans
+                ans = Range(A, B + _r.step, _r.step)
+                if r.step != ans.step:
+                    ans = ans.reversed
+                assert r.step == ans.step
+                assert result == ans, "%s.intersect(%s) == %s" % (r, i, ans)
```
Comment (smichr) [inline sympy/sets/tests/test_fancysets.py]: the `txt` is not right now

File: sympy/sets/fancysets.py (hunk 4 above)
Comment (smichr) [inline sympy/sets/fancysets.py]: oops - that fails

```
    assert len(Range(0, oo, 2)) == oo
  File "sympy\sympy\core\expr.py", line 199, in __int__
    raise TypeError("can't convert %s to int" % r)
TypeError: can't convert oo to int
```

File: sympy/sets/tests/test_fancysets.py (hunk 7 above)
Comment (smichr) [inline sympy/sets/tests/test_fancysets.py]: fuxed

**Comment by asmeurer** [comment]
If you're confident the checks are still OK without the explicit answer written out then this is fine. +1 to everything. I'm starting to think maybe you're right about returning Range(0, 0, 1) instead of EmptySet, but that can be a separate PR.
